matchzoo/preprocessors.py

```diff
@@ -0,0 +1,133 @@
+"""Matchzoo toolkit for text pre-processing."""
+
+import re
+import nltk
+import jieba
+import many_stop_words
+from functools import reduce
+
+
+def get_stopwords(lang: str='en') -> list:
+    """
+    Get stopwords based on language.
+
+    :params lang: language code.
+
+    :return stop_list: list of stop words.
+    """
+    return many_stop_words.get_stop_words(lang)
+
+
+def tokenizer(raw: str) -> list:
+    """
+    Word tokenization.
+
+    :param raw: raw input given by user.
+
+    :return tokens: list of tokenized tokens.
+    """
+    return nltk.word_tokenize(raw)
+
+
+def segmentation(raw: str) -> list:
+    """
+    Chinese sentence segmentation.
+
+    :param raw: raw input given by user.
+
+    :return tokens: list of segmentated tokens.
+    """
+    return list(jieba.cut(raw))
+
+
+def to_lowercase(tokens: list) -> list:
+    """
+    Convert list of tokens to lower case.
+
+    :param tokens: list of tokens.
+
+    :return tokens: lower-cased list of tokens.
+    """
+    return [token.lower() for token in tokens]
+
+
+def remove_stopwords(tokens: list, lang: str='en') -> list:
+    """
+    Remove stopwords from list of tokenized tokens.
+
+    :param tokens: list of tokenized tokens.
+    :param lang: language code for stopwords.
+
+    :return tokens: list of tokenized tokens without stopwords.
+    """
+    return [token for token in tokens if token not in get_stopwords(lang)]
+
+
+def remove_punctuation(tokens: list) -> list:
+    """
+    Remove punctuations from list of tokens.
+
+    :param tokens: list of
+
+    :return rv: tokens  without punctuation.
+    """
+    rv = []
+    match = re.compile('[^\w\s]')
```

Comment:
Well, the regex still compiles for each call here. Just move the compiled regex outside.

Reply:
updated